test/torch/tensors/test_autograd.py

```diff
@@ -511,3 +511,63 @@ def test_get_float_prec_on_autograd_tensor(workers):
     x2 = x.fix_precision()
     x2 = x2.share(bob, alice, crypto_provider=james, requires_grad=True)
     assert (x2.get().float_precision() == x).all()
+
+
+def test_serialize_deserialize_autograd_tensor(workers):
+    # let's try to send an autogradTensor to a remote location and get it back
+    alice = workers["alice"]
+    random_tensor = torch.randn(5, 3)
+    remote_tensor = random_tensor.send(alice, local_autograd=True)
+    local_tensor = remote_tensor.get()
+    # check if the tensor sent is equal to the tensor got from the remote version
+    equal_tensors = torch.all(torch.eq(local_tensor, random_tensor))
+
+    assert equal_tensors == 1
+
+
+def test_train_remote_autograd_tensor(workers):
+    def train(model_input, data_input, target_input):
+        opt = optim.SGD(params=model_input.parameters(), lr=0.1)
+        for iter in range(10):
+            # 1) erase previous gradients (if they exist)
+            opt.zero_grad()
+            # 2) make a prediction
+            predictions = model_input(data_input)
+            # 3) calculate how much we missed
+            loss = ((predictions - target_input) ** 2).sum()
+            # 4) Figure out which weights caused us to miss
+            loss.backward()
+            # 5) change those weights
+            opt.step()
+        return (loss, model_input)
+
+    # Training procedure to train an input model, be it remote or local
+    worker = workers["alice"]
+
+    # Some Toy Data
+    data_local = torch.tensor([[0, 0], [0, 1], [1, 0], [1, 1.0]])
+    target_local = torch.tensor([[0], [0], [1], [1.0]])
+
+    # Toy local model
+    model_local = nn.Linear(2, 1)
+
+    # Local training
+    loss_local, model_local_trained = train(model_local, data_local, target_local)
+
+    # Remote training, setting autograd for the data and the targets
+    data_remote = data_local.send(worker, local_autograd=True)
+    target_remote = target_local.send(worker, local_autograd=True)
+    model_remote = model_local.send(worker, local_autograd=True)
+
+    loss_remote, model_remote_trained = train(model_remote, data_remote, target_remote)
+
+    # let's check if the local version and the remote version have the same weigth
+    equal_weights = torch.all(
+        torch.eq(
+            model_local_trained.weight.copy().get().data, model_remote.weight.copy().get().data
+        )
+    )
+
+    assert equal_weights == 1
+
+    assert loss_remote.copy().get().item() < 500 and loss_local.item() < 500
```

Comment:
Let's see how we can improve this second assert statement.
A loss < 500 seems very lax to me. Does the fitting work on this dataset work with 10 epochs? I guess it does, given that the classifier just needs to look at the first coordinate (linearly separable problem, much easier than xor). 
If the training is stable, we could ask for both remote and local loss to decrease through training. 
I just saw that you send the local model after training to the remote location to train again. That makes it a bit more tricky to check the loss. Any propositions?

Reply:
Yeah, it is indeed a bit more tricky because it looks like the local model is being affected too, following the remote training. I tried to detach all the parameters of the local model, but that didn't make a difference.

As you suggested, we could probably check for a monotonic decrease of the loss during the training procedure both for the local and the remote training procedure.

Reply:
Agree.